cmd/xzcatomicswap/main.go

```diff
@@ -0,0 +1,1116 @@
+	default:
+		return ""
+	}
+}
+
+// createSig creates and returns the serialized raw signature and compressed
+// pubkey for a transaction input signature.  Due to limitations of the Zcoin
+// Core RPC API, this requires dumping a private key and signing in the client,
+// rather than letting the wallet sign.
+func createSig(tx *wire.MsgTx, idx int, pkScript []byte, addr xzcutil.Address,
+	c *rpc.Client) (sig, pubkey []byte, err error) {
+
+	wif, err := c.DumpPrivKey(addr)
+	if err != nil {
+		return nil, nil, err
+	}
+	sig, err = txscript.RawTxInSignature(tx, idx, pkScript, txscript.SigHashAll, wif.PrivKey)
+	if err != nil {
+		return nil, nil, err
+	}
+	return sig, wif.PrivKey.PubKey().SerializeCompressed(), nil
+}
+
+// fundRawTransaction calls the fundrawtransaction JSON-RPC method.  It is
+// implemented manually as client support is currently missing from the
+// xzcd/rpcclient package.
+func fundRawTransaction(c *rpc.Client, tx *wire.MsgTx, feePerKb xzcutil.Amount) (fundedTx *wire.MsgTx, fee xzcutil.Amount, err error) {
+	var buf bytes.Buffer
+	buf.Grow(tx.SerializeSize())
+	tx.Serialize(&buf)
+	param0, err := json.Marshal(hex.EncodeToString(buf.Bytes()))
+	if err != nil {
+		return nil, 0, err
+	}
+	param1, err := json.Marshal(struct {
+		FeeRate float64 `json:"feeRate"`
+	}{
+		FeeRate: feePerKb.ToXZC(),
+	})
+	if err != nil {
+		return nil, 0, err
+	}
+	params := []json.RawMessage{param0, param1}
+	rawResp, err := c.RawRequest("fundrawtransaction", params)
+	if err != nil {
+		return nil, 0, err
+	}
+	var resp struct {
+		Hex       string  `json:"hex"`
+		Fee       float64 `json:"fee"`
+		ChangePos float64 `json:"changepos"`
+	}
+	err = json.Unmarshal(rawResp, &resp)
+	if err != nil {
+		return nil, 0, err
+	}
+	fundedTxBytes, err := hex.DecodeString(resp.Hex)
+	if err != nil {
+		return nil, 0, err
+	}
+	fundedTx = &wire.MsgTx{}
+	err = fundedTx.Deserialize(bytes.NewReader(fundedTxBytes))
+	if err != nil {
+		return nil, 0, err
+	}
+	feeAmount, err := xzcutil.NewAmount(resp.Fee)
+	if err != nil {
+		return nil, 0, err
+	}
+	return fundedTx, feeAmount, nil
+}
+
+// getFeePerKb queries the wallet for the transaction relay fee/kB to use and
+// the minimum mempool relay fee.  It first tries to get the user-set fee in the
+// wallet.  If unset, it attempts to find an estimate using estimatefee 6.  If
+// both of these fail, it falls back to mempool relay fee policy.
+func getFeePerKb(c *rpc.Client) (useFee, relayFee xzcutil.Amount, err error) {
+	info, err := c.GetInfo()
+	if err != nil {
+		return 0, 0, fmt.Errorf("getinfo: %v", err)
+	}
+	relayFee, err = xzcutil.NewAmount(info.RelayFee)
+	if err != nil {
+		return 0, 0, err
+	}
+	maxFee := info.PaytxFee
+	if info.PaytxFee != 0 {
+		if info.RelayFee > maxFee {
+			maxFee = info.RelayFee
+		}
+		useFee, err = xzcutil.NewAmount(maxFee)
+		return useFee, relayFee, err
+	}
+
+	params := []json.RawMessage{[]byte("6")}
+	estimateRawResp, err := c.RawRequest("estimatefee", params)
```

Comment:
I realize that we just got estimatefee working after your zcoin PR, but it also appears that it implements estimatesmartfee which is the replacement for estimatefee in recent Bitcoin Core versions.  It would probably be wise to update to use that.  See the recent changes to the other tools which update the getinfo call to getnetworkinfo+getwalletinfo, and estimatefee to estimatesmartfee.

Reply:
`estimatesmartfee` is also implemented in Zcoin - (bitcoin 0.13 based) and I could put in a separate PR after merging this PR to update **xzcatomicswap** code to use `estimatesmartfee` RPC. 
 I did note your updates to btcd. In terms of updating bitcoin core I am not at liberty to comment at this time but would expect it not to be updated to 0.15 or 0.16 anytime soon.